### PR TITLE
Add native version of black scholes algo as an example

### PR DESCRIPTION
**Pull request by densmirn**

### CODE REVIEW

File: dpnp/backend/examples/example_bs.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    double* d1 = (double*)dpnp_memory_alloc_c(size * sizeof(double));
+    dpnp_divide_c<double, double, double>(S, K, d1, size);         // S/K
+    dpnp_log_c<double, double>(d1, d1, size);               // np.log(S/K)
+
+    double* r_sigma_sigma_2_t = (double*)dpnp_memory_alloc_c(size * sizeof(double));
+    dpnp_multiply_c<double, double, double>(r_sigma_sigma_2, T, r_sigma_sigma_2_t, size);  // r_sigma_sigma_2*T
+    dpnp_add_c<double, double, double>(d1, r_sigma_sigma_2_t, d1, size);     // np.log(S/K) + r_sigma_sigma_2*T
+    dpnp_memory_free_c(r_sigma_sigma_2_t);
+
+    double* sigmas_sqrtt = (double*)dpnp_memory_alloc_c(size * sizeof(double));
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:54]: You are allocating mem on 49 and then deallocating it on 52 line, after you allocating the same memory on 54th line. Please do not make multiple allocations on your code, just reuse once allocated mem.

File: dpnp/backend/examples/example_bs.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+/**
+ * Example 3.
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:27]: check the name, please

File: dpnp/backend/examples/example_bs.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+    // K*np.exp(nrs*T)*cdf_d2
+    dpnp_multiply_c<double, double, double>(temp, cdf_d2, temp, size);
+    dpnp_memory_free_c(cdf_d2);
+
+    // S*cdf_d1 - K*np.exp(nrs*T)*cdf_d2
+    dpnp_subtract_c<double, double, double>(bs_call, temp, bs_call, size);
+    dpnp_memory_free_c(temp);
+
+    double* bs_put = (double*)dpnp_memory_alloc_c(size * sizeof(double));
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:92]: reuse `temp` here please.

File: dpnp/backend/examples/example_bs.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    const double RISK_FREE = 0.1;
+    const double VOLATILITY = 0.2;
+
+    dpnp_queue_initialize_c(QueueOptions::GPU_SELECTOR);
+    std::cout << "SYCL queue is CPU: " << dpnp_queue_is_cpu_c() << std::endl;
+
+    double* S = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
+    double* K = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
+    double* T = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
+    double* result = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:118]: You don't use this allocated memory at all. please, remove it.

File: dpnp/backend/examples/example_bs.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    nr[0] = -RISK_FREE;
+
+    double* sq2 = (double*)dpnp_memory_alloc_c(1 * sizeof(double));
+    sq2[0] = sqrt(2.);
+
+    double* one = (double*)dpnp_memory_alloc_c(1 * sizeof(double));
+    one[0] = 1.;
+
+    double* two = (double*)dpnp_memory_alloc_c(1 * sizeof(double));
+    two[0] = 2.;
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:143]:
```suggestion
double* params_for_full = (double*)dpnp_memory_alloc_c(7 * sizeof(double));
params_for_full[0] = RISK_FREE;
params_for_full[1] = VOLATILITY;
params_for_full[2] = RISK_FREE + VOLATILITY*VOLATILITY/2.;
params_for_full[3] = -RISK_FREE;
params_for_full[4] = sqrt(2.);
params_for_full[5] = 1.;
params_for_full[6] = 2.;
```
update on other places and just free `params_for_full` after use.

Please, again, don't do multiple call of mem allocator. This is bad code. Use at once allocated mem and don't call allocator multiple times as much as possible.

File: dpnp/backend/examples/example_bs.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    const long TL = 1,  TH = 2;
+    const double RISK_FREE = 0.1;
+    const double VOLATILITY = 0.2;
+
+    dpnp_queue_initialize_c(QueueOptions::GPU_SELECTOR);
+    std::cout << "SYCL queue is CPU: " << dpnp_queue_is_cpu_c() << std::endl;
+
+    double* S = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
+    double* K = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
+    double* T = (double*)dpnp_memory_alloc_c(SIZE * sizeof(double));
```
Comment (samir-nasibli) [inline dpnp/backend/examples/example_bs.cpp:117]:
```suggestion
double* S = (double*)dpnp_memory_alloc_c(SIZE * 3 * sizeof(double));
double* K = S + SIZE;
double* T = K + SIZE;
```
This is better.

File: dpnp/backend/examples/example_bs.cpp (hunk 5 above)
Comment (densmirn) [reply dpnp/backend/examples/example_bs.cpp:143]: I agree, but the example won't be comparable with python version with the change.

File: dpnp/backend/examples/example_bs.cpp (hunk 6 above)
Comment (densmirn) [reply dpnp/backend/examples/example_bs.cpp:117]: I agree, but the example won't be comparable with python version with the change.

File: dpnp/backend/examples/example_bs.cpp (hunk 5 above)
Comment (samir-nasibli) [reply dpnp/backend/examples/example_bs.cpp:145]: Do you think that python manages mem as you presented? It is disputable, I think.

File: dpnp/backend/examples/example_bs.cpp (hunk 6 above)
Comment (samir-nasibli) [reply dpnp/backend/examples/example_bs.cpp:119]: Do you think that python manages mem as you presented? It is disputable.

Comment (samir-nasibli) [reply dpnp/backend/examples/example_bs.cpp:119]: And also there is no description about comparison with python version or `python-like` impl code. Please update description.

File: dpnp/backend/examples/example_bs.cpp (hunk 5 above)
Comment (densmirn) [reply dpnp/backend/examples/example_bs.cpp:145]: We manage memory the same way in Cython level of the function `full`.

Comment (samir-nasibli) [reply dpnp/backend/examples/example_bs.cpp:145]: Again this is bad case in any way. Python and Cython differs on using mem in some cases be aware. You need to look Cython generation properly for this place of you cython written code.

Also there is no given description about comparison with python version or python-like impl code. Please update the description on src and for PR as well.

**Review by shssf — APPROVED** [review]
Examples must be as clear as possible. No complex or algorithm optimizations are required.

**Pull request merged**